Mathlib/Tactic/Linter/Lint.lean

```diff
@@ -293,6 +293,80 @@ initialize addLinter lambdaSyntaxLinter
 
 end Style.lambdaSyntax
 
+/-!
+#  The "longFile" linter
+
+The "longFile" linter emits a warning on files which are longer than a certain number of lines
+(1500 by default).
+-/
+
+/--
+The "longFile" linter emits a warning on files which are longer than a certain number of lines
+(1500 by default). If this option is set to `N` lines, the linter warns once a file has more than
+`N` lines. A value of `0` silences the linter entirely.
+-/
+register_option linter.style.longFile : Nat := {
+  defValue := 1500
+  descr := "enable the longFile linter"
+}
+
+namespace Style.longFile
+
+@[inherit_doc Mathlib.Linter.linter.style.longFile]
+def longFileLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  let linterBound := linter.style.longFile.get (← getOptions)
+  if linterBound == 0 then
+    return
+  let defValue := linter.style.longFile.defValue
+  let smallOption := match stx with
+      | `(set_option linter.style.longFile $x) => TSyntax.getNat ⟨x.raw⟩ ≤ defValue
+      | _ => false
+  if smallOption then
+    logWarningAt stx <| .tagged linter.style.longFile.name
+      m!"The default value of the `longFile` linter is {defValue}.\n\
+        The current value of {linterBound} does not exceed the allowed bound.\n\
+        Please, remove the `set_option linter.style.longFile {linterBound}`."
+  else
+  -- Thanks to the above check, the linter option is either not set (and hence equal
+  -- to the default) or set to some value *larger* than the default.
+  -- `Parser.isTerminalCommand` allows `stx` to be `#exit`: this is useful for tests.
+  unless Parser.isTerminalCommand stx do return
```

Comment:
Great that you found a way to test this! Let's add a comment about user implications: how about the following?
```
-- (This also means the linter will stop at an #exit command: if the user sets this linter option *before* the #exit, this could generate slightly confusing warnings if the linter suggests lowering the value.
-- Hopefully, in practice the linter is only changed at the end of the file, so this will not matter.)
```

Reply:
I do not understand what this comment is clarifying: this is probably due to the fact that I have thought quite a bit about the linter and I am failing to see what can be confusing about it!

Can you explain what confusion this comment alleviates?  :smile:

Reply:
I worried about the following scenario:
- in a long file, the set_option is set "early" (say, at the beginning of the file)
- the user temporarily writes #exit, say, to debug something
This exit is placed *after* the set_option, but early enough that *at this point*, the linter option is set too high.

Then, the linter will complain about the option being set too high --- but that complain is spurious (it will disappear once the #exit is removed) and following the warning will yield *another* (correct, non-spurious) warning when the #exit is removed again.

Granted, this scenario is a bit convoluted (and only a temporary state).
Given this, your suggestion of warning about the option being set early seems useful (if perf. allows it; otherwise, a comment should suffice).

Reply:
I see.  Honestly, if someone knows that `#exit` shielded inside a `xxx in #exit` allows Lean to continue parsing the file after `#exit`, I think that they will not be confused by the linter for too long!  I would leave this as is, though I can implement the forcing the `set_option` to be the "last" command of a file.

Reply:
Sure - this is not something I consider essential. While minimising surprises is nice and these things add up, there are probably more useful changes one could make in that regard.